src/mumble/JackAudio.cpp

```diff
@@ -0,0 +1,618 @@
+		qWarning("JackAudioSystem: unable to open jack client due to %i errors:", errors.count());
+		for (int i = 0; i < errors.count(); ++i) {
+			qWarning("JackAudioSystem:  %s", qPrintable(errors.at(i)));
+		}
+		bJackIsGood = false;
+		client = NULL;
+	}
+}
+
+void JackAudioSystem::close_jack() {
+
+	QMutexLocker lock(&qmWait);
+	if (client) {
+		int err = 0;
+		err = jack_deactivate(client);
+		if (err != 0)  {
+			qWarning("JackAudioSystem: unable to remove client from the process graph - jack_deactivate() returned %i", err);
+		}
+
+		bActive = false;
+
+		if (in_port != NULL) {
+			err = jack_port_unregister(client, in_port);
+			if (err != 0)  {
+				qWarning("JackAudioSystem: unable to unregister in port - jack_port_unregister() returned %i", err);
+			}
+		}
+
+		for (unsigned i = 0; i < iOutPorts; ++i) {
+			if (out_ports[i] != NULL) {
+				err = jack_port_unregister(client, out_ports[i]);
+				if (err != 0)  {
+					qWarning("JackAudioSystem: unable to unregister out port - jack_port_unregister() returned %i", err);
+				}
+			}
+		}
+
+		err = jack_client_close(client);
+		if (err != 0) {
+			qWarning("JackAudioSystem: unable to to disconnect from the JACK server - jack_client_close() returned %i", err);
+		}
+
+		delete [] output_buffer;
+		output_buffer = NULL;
+
+		client = NULL;
+	}
+	bJackIsGood = false;
+}
+
+
+void JackAudioSystem::auto_connect_ports()
+{
+	if (g.s.bJackAutoConnect == false) {
+		return;
+	}
+
+	const char **ports = NULL;
+	const int wanted_out_flags = JackPortIsPhysical | JackPortIsOutput;
+	const int wanted_in_flags = JackPortIsPhysical | JackPortIsInput;
+	int err;
+	unsigned int connected_out_ports = 0;
+	unsigned int connected_in_ports = 0;
+
+	ports = jack_get_ports(client, 0, "audio", JackPortIsPhysical);
+	if (ports != NULL) {
+		int i = 0;
+		while (ports[i] != NULL) {
+			jack_port_t * const port = jack_port_by_name(client, ports[i]);
+			if (port == NULL)  {
+				qWarning("JackAudioSystem: jack_port_by_name() returned an invalid port - skipping it");
+				continue;
+			}
+
+			const int port_flags = jack_port_flags(port);
+
+			if ((port_flags & wanted_out_flags) == wanted_out_flags && connected_in_ports < 1) {
+				err = jack_connect(client, ports[i], jack_port_name(in_port));
+				if (err != 0) {
+					qWarning("JackAudioSystem: unable to connect port '%s' to '%s' - jack_connect() returned %i", ports[i], jack_port_name(in_port), err);
+				} else {
+					connected_in_ports++;
+				}
+			}
+			else if ((port_flags & wanted_in_flags) == wanted_in_flags && connected_out_ports < iOutPorts) {
+				err = jack_connect(client, jack_port_name(out_ports[connected_out_ports]), ports[i]);
+				if (err != 0) {
+					qWarning("JackAudioSystem: unable to connect port '%s' to '%s' - jack_connect() returned %i", jack_port_name(out_ports[connected_out_ports]), ports[i], err);
+				} else {
+					connected_out_ports++;
+				}
+			}
+
+			++i;
+		}
+	}
+}
+
+void JackAudioSystem::activate()
+{
```

Comment:
The bracket should be at the end of the line above (to match the style across the file).

Reply:
Done